tracetools/CMakeLists.txt

```
@@ -80,6 +80,10 @@ if(TRACETOOLS_LTTNG_ENABLED)
   if(NOT TRACETOOLS_NO_RDYNAMIC)
     target_link_libraries(${PROJECT_NAME} "-rdynamic")
   endif()
+elseif(NOT WIN32)
+  # Explicitly linking against dl for dladdr() seems to be required for RHEL
+  # (LTTng's link flags already include it above)
+  target_link_libraries(${PROJECT_NAME} "-ldl")
 endif()
```

Thanks for finding a fix so quickly. I'm happy to see that this didn't take out the Ubuntu builds like I initially thought it did.

This is probably fine as-is, but I have two soft suggestions:
1. This is an opportunity to employ a more defensive programming technique with regard to external dependencies changing their projects. I'm not sure it's a good idea to rely on `lttng-ust` to provide a library that's a direct dependency of this one.
2. Use the `CMAKE_LINK_LIBRARY_FLAG` macro instead of verbatim `-l`.

```suggestion
if(NOT WIN32)
  # Explicitly linking against dl for dladdr() seems to be required for RHEL
  # (LTTng's link flags already include it above)
  target_link_libraries(${PROJECT_NAME} "${CMAKE_LINK_LIBRARY_FLAG}dl")
endif()
```

---

Another thing to note here is that the reason RHEL hit this is that we're not installing the `lttng-ust` development package in our CI at all. I took a look, and it it would have been present, it would have provided the `-ldl` link flag. Interestingly, that dependency isn't declared in this package's manifest.

The reason we're not installing `lttng-ust-devel` is that the system package version is too old for this package to use. I vaguely remember opening an issue or making a PR a while ago to improve the dependency version detection to reject the older package, but I can't recall where that landed.

---

Another interesting note is that our nightly CI builds on [build.ros2.org](https://build.ros2.org/view/Rci/job/Rci__nightly-release_ubuntu_jammy_amd64/369) install dependencies using rosdep and therefore didn't install the `lttng-ust` development package, but it didn't cause the same problem. Maybe Ubuntu links in `dl` implicitly?

---

> This is probably fine as-is, but I have two soft suggestions:

Thanks!

> Another thing to note here is that the reason RHEL hit this is that we're not installing the `lttng-ust` development package in our CI at all. I took a look, and it it would have been present, it would have provided the `-ldl` link flag.

That's correct, but the reason this broke after #43 is because we were `#ifdef`ing the `dladdr()` call out if not `TRACETOOLS_LTTNG_ENABLED`.

> Interestingly, that dependency isn't declared in this package's manifest.
> The reason we're not installing lttng-ust-devel is that the system package version is too old for this package to use. I vaguely remember opening an issue or making a PR a while ago to improve the dependency version detection to reject the older package, but I can't recall where that landed.

I think it was this issue: https://github.com/ros2/ros2_tracing/issues/16 (or the original GitLab issue). However, I think you recently said that RHEL has LTTng 2.12, which would be fine (unlike 2.8 like you mentioned in the original GitLab issue).

We intentionally do not declare a dependency on LTTng-UST, but we plan on changing that soon, see https://github.com/ros2/ros2/issues/1177.

---

> Maybe Ubuntu links in dl implicitly?

Yeah, it would seem so.

---

> I think you recently said that RHEL has LTTng 2.12

Yes, RHEL 9 has LTTng 2.12, but we haven't quite switched Rolling over from RHEL 8 yet, which has 2.8.1. It would probably be a good idea to follow through with #16 prior to deploying ros2/ros2#1177 even after we get Rolling onto RHEL 9 - I think it's safe to say that older versions are not uncommon in the wild.